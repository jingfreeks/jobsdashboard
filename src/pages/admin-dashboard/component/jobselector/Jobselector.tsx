 
import {
  Headers,
  SearchFilter,
  JobsList,
  AddModal,
  EditModal,
  Loading,
  Error
} from "./component";
import {useJobSelectorHooks} from "./hooks";
const JobSelector = () => {
  const hooks = useJobSelectorHooks();
  // Loading state
  if (hooks.isLoading) {
    return (
      <Loading />
    );
  }

  // Error state
  if (hooks.error) {
    return (
      <Error />
    );
  }

  return (
    <div className="max-w-6xl mx-auto">
      <div className="fixed top-4 right-4 z-50 space-y-2">
        {/* Toast notifications will appear here */}
      </div>

      <Headers handleAddJob={hooks.handleAddJob} />
      {/* Search and Filter Section */}
      <SearchFilter
        jobsWithDetails={hooks.jobsWithDetails}
        companies={hooks.companies}
        cities={hooks.cities}
        departments={hooks.departments}
        selectedCompanyFilter={hooks.selectedCompanyFilter}
        setSelectedCompanyFilter={hooks.setSelectedCompanyFilter}
        selectedCityFilter={hooks.selectedCityFilter}
        setSelectedCityFilter={hooks.setSelectedCityFilter}
        selectedDepartmentFilter={hooks.selectedDepartmentFilter}
        setSelectedDepartmentFilter={hooks.setSelectedDepartmentFilter}
        searchTerm={hooks.searchTerm}
        setSearchTerm={hooks.setSearchTerm}
        filteredJobs={hooks.filteredJobs}
      />
      {/* Jobs List */}
      <JobsList
        filteredJobs={hooks.filteredJobs}
        handleEditJob={hooks.handleEditJob}
        handleDeleteJob={hooks.handleDeleteJob}
        isDeleting={hooks.isDeleting}
        jobsWithDetails={hooks.jobsWithDetails}
      />
      {/* Add Job Modal */}
      {hooks.showAddJobModal && (
        <AddModal
          isAdding={hooks.isAdding}
          handleCloseAddJobModal={hooks.handleCloseAddJobModal}
          handleCreateJob={hooks.handleCreateJob}
          companies={hooks.companies}
          cities={hooks.cities}
          departments={hooks.departments}
        />
      )}

      {/* Edit Job Modal */}
      {hooks.showEditJobModal && hooks.editJob && (
        <EditModal
          handleCloseEditJobModal={hooks.handleCloseEditJobModal}
          handleUpdateJob={hooks.handleUpdateJob}
          isUpdating={hooks.isUpdating}
          editJob={hooks.editJob}
          companies={hooks.companies}
          cities={hooks.cities}
          departments={hooks.departments}
        />
      )}
    </div>
  );
};

export default JobSelector;
