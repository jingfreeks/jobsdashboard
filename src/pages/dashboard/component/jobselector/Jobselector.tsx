import { PlusCircle } from "lucide-react";

type JobsTypes = {
  handleAddJob: () => void;
  showAddJobModal: boolean;
  handleCloseModal: () => void;
  handleCreateJob: (e: React.FormEvent) => void;
  newJobTitle: string;
  jobs: { id: string; title: string }[];
  setNewJobTitle: (title: string) => void;
  handleEditJob: (id: string) => void;
  showEditJobModal: boolean;
  handleCloseEditModal: () => void;
  handleUpdateJob: (e: React.FormEvent) => void;
  handleDeleteJob: (id: string) => void;
  editJobTitle: string;
  setEditJobTitle: (title: string) => void;
};

const Jobselector = (props:JobsTypes) => {
  const {
    handleAddJob,
    showAddJobModal,
    handleCloseModal,
    handleCreateJob,
    newJobTitle,
    jobs,
    setNewJobTitle,
    handleEditJob,
    showEditJobModal,
    handleCloseEditModal,
    handleUpdateJob,
    handleDeleteJob,
    editJobTitle,
    setEditJobTitle,
  } = props;
  return (
    <div className="max-w-2xl mx-auto">
      <div className="flex items-center justify-between mb-6">
        <h2 className="text-2xl font-bold text-slate-800">Jobs List</h2>
        <button
          onClick={handleAddJob}
          className="flex items-center gap-2 bg-blue-600 hover:bg-blue-700 text-white px-4 py-2 rounded-lg font-semibold shadow transition"
        >
          <PlusCircle className="w-5 h-5" /> Add Job
        </button>
      </div>
      {/* Add Job Modal */}
      {showAddJobModal && (
        <div className="fixed inset-0 z-50 flex items-center justify-center bg-black bg-opacity-40">
          <div className="bg-white rounded-xl shadow-lg p-8 w-full max-w-md relative">
            <button
              onClick={handleCloseModal}
              className="absolute top-3 right-3 text-slate-400 hover:text-slate-700 text-xl font-bold"
            >
              &times;
            </button>
            <h3 className="text-xl font-bold mb-4 text-slate-800">
              Create New Job
            </h3>
            <form onSubmit={handleCreateJob} className="flex flex-col gap-4">
              <input
                type="text"
                className="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
                placeholder="Job Title"
                value={newJobTitle}
                onChange={(e) => setNewJobTitle(e.target.value)}
                autoFocus
                required
              />
              <div className="flex gap-2 justify-end">
                <button
                  type="button"
                  onClick={handleCloseModal}
                  className="px-4 py-2 rounded-lg bg-slate-100 hover:bg-slate-200 text-slate-700 font-semibold"
                >
                  Cancel
                </button>
                <button
                  type="submit"
                  className="px-4 py-2 rounded-lg bg-blue-600 hover:bg-blue-700 text-white font-semibold"
                >
                  Create
                </button>
              </div>
            </form>
          </div>
        </div>
      )}
      <div className="bg-white rounded-xl shadow p-6 border border-slate-100">
        {jobs.length === 0 ? (
          <div className="text-slate-400 italic">No jobs</div>
        ) : (
          <ul className="divide-y divide-slate-100">
            {jobs.map((job) => (
              <li
                key={job.id}
                className="flex items-center justify-between py-3"
              >
                <span className="flex-1 truncate text-slate-800 font-medium">
                  {job.title}
                </span>
                <div className="flex gap-2">
                  <button
                    onClick={() => handleEditJob(job.id)}
                    className="text-blue-500 hover:text-blue-700 px-2 py-1 rounded transition"
                  >
                    <svg
                      xmlns="http://www.w3.org/2000/svg"
                      className="h-4 w-4"
                      fill="none"
                      viewBox="0 0 24 24"
                      stroke="currentColor"
                    >
                      <path
                        strokeLinecap="round"
                        strokeLinejoin="round"
                        strokeWidth={2}
                        d="M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z"
                      />
                    </svg>
                  </button>
                  <button
                    onClick={() => handleDeleteJob(job.id)}
                    className="text-red-500 hover:text-red-700 px-2 py-1 rounded transition"
                  >
                    <svg
                      xmlns="http://www.w3.org/2000/svg"
                      className="h-4 w-4"
                      fill="none"
                      viewBox="0 0 24 24"
                      stroke="currentColor"
                    >
                      <path
                        strokeLinecap="round"
                        strokeLinejoin="round"
                        strokeWidth={2}
                        d="M6 18L18 6M6 6l12 12"
                      />
                    </svg>
                  </button>
                </div>
              </li>
            ))}
          </ul>
        )}
      </div>
      {/* Edit Job Modal */}
      {showEditJobModal && (
        <div className="fixed inset-0 z-50 flex items-center justify-center bg-black bg-opacity-40">
          <div className="bg-white rounded-xl shadow-lg p-8 w-full max-w-md relative">
            <button
              onClick={handleCloseEditModal}
              className="absolute top-3 right-3 text-slate-400 hover:text-slate-700 text-xl font-bold"
            >
              &times;
            </button>
            <h3 className="text-xl font-bold mb-4 text-slate-800">Edit Job</h3>
            <form onSubmit={handleUpdateJob} className="flex flex-col gap-4">
              <input
                type="text"
                className="border border-slate-200 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-500"
                placeholder="Job Title"
                value={editJobTitle}
                onChange={(e) => setEditJobTitle(e.target.value)}
                autoFocus
                required
              />
              <div className="flex gap-2 justify-end">
                <button
                  type="button"
                  onClick={handleCloseEditModal}
                  className="px-4 py-2 rounded-lg bg-slate-100 hover:bg-slate-200 text-slate-700 font-semibold"
                >
                  Cancel
                </button>
                <button
                  type="submit"
                  className="px-4 py-2 rounded-lg bg-blue-600 hover:bg-blue-700 text-white font-semibold"
                >
                  Update
                </button>
              </div>
            </form>
          </div>
        </div>
      )}
    </div>
  );
};
export default Jobselector;
