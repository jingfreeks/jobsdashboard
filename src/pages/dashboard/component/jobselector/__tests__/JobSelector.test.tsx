import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import Jobselector from '../Jobselector';

describe('JobSelector', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should render correctly and match snapshot', () => {
    const { container } = render(<Jobselector />);
    expect(container).toMatchSnapshot();
  });

  it('should display jobs list correctly', () => {
    render(<Jobselector />);

    expect(screen.getByText('Jobs List')).toBeInTheDocument();
    expect(screen.getByText('Frontend Developer')).toBeInTheDocument();
    expect(screen.getByText('Backend Engineer')).toBeInTheDocument();
    expect(screen.getByText('Product Manager')).toBeInTheDocument();
  });

  it('should show empty state when no jobs exist', () => {
    // We need to modify the component to start with empty jobs for this test
    // Since the component has hardcoded initial jobs, we'll test the empty state
    // by deleting all jobs first
    render(<Jobselector />);
    
    // Delete all jobs
    const deleteButtons = screen.getAllByRole('button').filter(button => 
      button.innerHTML.includes('M6 18L18 6M6 6l12 12')
    );
    
    deleteButtons.forEach(button => {
      fireEvent.click(button);
    });

    expect(screen.getByText('No jobs')).toBeInTheDocument();
  });

  it('should open add job modal when add button is clicked', () => {
    render(<Jobselector />);
    
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    expect(screen.getByText('Create New Job')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Job Title')).toBeInTheDocument();
  });

  it('should create a new job successfully', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Fill form and submit
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: 'New Job Title' } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(screen.getByText('New Job Title')).toBeInTheDocument();
    });
  });

  it('should not create job with empty name', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Try to submit empty form
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    // Modal should still be open
    expect(screen.getByText('Create New Job')).toBeInTheDocument();
  });

  it('should not create job with whitespace-only name', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Fill form with whitespace and submit
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: '   ' } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    // Modal should still be open
    expect(screen.getByText('Create New Job')).toBeInTheDocument();
  });

  it('should trim whitespace from job title when creating', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Fill form with whitespace and submit
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: '  New Job Title  ' } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(screen.getByText('New Job Title')).toBeInTheDocument();
    });
  });

  it('should open edit modal when edit button is clicked', () => {
    render(<Jobselector />);
    
    // Find and click edit button for the first job
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      expect(screen.getByText('Edit Job')).toBeInTheDocument();
      expect(screen.getByDisplayValue('Frontend Developer')).toBeInTheDocument();
    }
  });

  it('should update job successfully', async () => {
    render(<Jobselector />);
    
    // Find and click edit button
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      
      // Update the title
      const input = screen.getByDisplayValue('Frontend Developer');
      fireEvent.change(input, { target: { value: 'Updated Job Title' } });
      
      const updateButton = screen.getByText('Update');
      fireEvent.click(updateButton);

      await waitFor(() => {
        expect(screen.getByText('Updated Job Title')).toBeInTheDocument();
      });
    }
  });

  it('should not update job with empty name', async () => {
    render(<Jobselector />);
    
    // Find and click edit button
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      
      // Clear the title
      const input = screen.getByDisplayValue('Frontend Developer');
      fireEvent.change(input, { target: { value: '' } });
      
      const updateButton = screen.getByText('Update');
      fireEvent.click(updateButton);

      // Modal should still be open
      expect(screen.getByText('Edit Job')).toBeInTheDocument();
    }
  });

  it('should trim whitespace from job title when updating', async () => {
    render(<Jobselector />);
    
    // Find and click edit button
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      
      // Update with whitespace
      const input = screen.getByDisplayValue('Frontend Developer');
      fireEvent.change(input, { target: { value: '  Updated Job Title  ' } });
      
      const updateButton = screen.getByText('Update');
      fireEvent.click(updateButton);

      await waitFor(() => {
        expect(screen.getByText('Updated Job Title')).toBeInTheDocument();
      });
    }
  });

  it('should delete job when delete button is clicked', async () => {
    render(<Jobselector />);
    
    // Verify initial job exists
    expect(screen.getByText('Frontend Developer')).toBeInTheDocument();
    
    // Find and click delete button
    const deleteButtons = screen.getAllByRole('button');
    const deleteButton = deleteButtons.find(button => 
      button.innerHTML.includes('M6 18L18 6M6 6l12 12')
    );
    
    if (deleteButton) {
      fireEvent.click(deleteButton);

      await waitFor(() => {
        expect(screen.queryByText('Frontend Developer')).not.toBeInTheDocument();
      });
    }
  });

  it('should close add modal when cancel button is clicked', () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Click cancel button
    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);

    expect(screen.queryByText('Create New Job')).not.toBeInTheDocument();
  });

  it('should close add modal when close button (×) is clicked', () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Click close button
    const closeButton = screen.getByText('×');
    fireEvent.click(closeButton);

    expect(screen.queryByText('Create New Job')).not.toBeInTheDocument();
  });

  it('should close edit modal when cancel button is clicked', () => {
    render(<Jobselector />);
    
    // Open edit modal
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      
      // Click cancel button
      const cancelButton = screen.getByText('Cancel');
      fireEvent.click(cancelButton);

      expect(screen.queryByText('Edit Job')).not.toBeInTheDocument();
    }
  });

  it('should close edit modal when close button (×) is clicked', () => {
    render(<Jobselector />);
    
    // Open edit modal
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      
      // Click close button
      const closeButton = screen.getByText('×');
      fireEvent.click(closeButton);

      expect(screen.queryByText('Edit Job')).not.toBeInTheDocument();
    }
  });

  it('should generate correct IDs for new jobs', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Add first new job
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: 'First New Job' } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(screen.getByText('First New Job')).toBeInTheDocument();
    });

    // Add second new job
    fireEvent.click(addButton);
    const input2 = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input2, { target: { value: 'Second New Job' } });
    
    const createButton2 = screen.getByText('Create');
    fireEvent.click(createButton2);

    await waitFor(() => {
      expect(screen.getByText('Second New Job')).toBeInTheDocument();
    });
  });

  it('should handle multiple job operations correctly', async () => {
    render(<Jobselector />);
    
    // Verify initial jobs
    expect(screen.getByText('Frontend Developer')).toBeInTheDocument();
    expect(screen.getByText('Backend Engineer')).toBeInTheDocument();
    expect(screen.getByText('Product Manager')).toBeInTheDocument();

    // Add a new job
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);
    
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: 'New Job' } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(screen.getByText('New Job')).toBeInTheDocument();
    });
  });

  it('should handle rapid job operations', async () => {
    render(<Jobselector />);
    
    // Open add modal multiple times quickly
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);
    fireEvent.click(addButton);
    fireEvent.click(addButton);

    // Should only show one modal
    const modals = screen.getAllByText('Create New Job');
    expect(modals).toHaveLength(1);
  });

  it('should handle special characters in job titles', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Add job with special characters
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: 'Job Title with @#$%^&*()' } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(screen.getByText('Job Title with @#$%^&*()')).toBeInTheDocument();
    });
  });

  it('should handle long job titles', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Add job with long title
    const longTitle = 'This is a very long job title that should be handled properly by the component without breaking the layout or functionality';
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: longTitle } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(screen.getByText(longTitle)).toBeInTheDocument();
    });
  });

  it('should maintain job order after operations', async () => {
    render(<Jobselector />);
    
    // Verify initial order
    const initialJobs = ['Frontend Developer', 'Backend Engineer', 'Product Manager'];
    initialJobs.forEach(job => {
      expect(screen.getByText(job)).toBeInTheDocument();
    });

    // Add a new job
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);
    
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: 'New Job' } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(screen.getByText('New Job')).toBeInTheDocument();
    });
  });

  it('should handle form submission with Enter key', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Fill form and submit with Enter key
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: 'Enter Key Job' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });

    // The component doesn't handle Enter key submission, so modal should remain open
    expect(screen.getByText('Create New Job')).toBeInTheDocument();
  });

  it('should handle edit form submission with Enter key', async () => {
    render(<Jobselector />);
    
    // Open edit modal
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      
      // Update with Enter key
      const input = screen.getByDisplayValue('Frontend Developer');
      fireEvent.change(input, { target: { value: 'Enter Key Update' } });
      fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });

      // The component doesn't handle Enter key submission, so modal should remain open
      expect(screen.getByText('Edit Job')).toBeInTheDocument();
    }
  });

  it('should handle empty state after deleting all jobs', async () => {
    render(<Jobselector />);
    
    // Delete all jobs
    const deleteButtons = screen.getAllByRole('button').filter(button => 
      button.innerHTML.includes('M6 18L18 6M6 6l12 12')
    );
    
    deleteButtons.forEach(button => {
      fireEvent.click(button);
    });

    await waitFor(() => {
      expect(screen.getByText('No jobs')).toBeInTheDocument();
    });

    // Verify add button is still available
    expect(screen.getByText('Add Job')).toBeInTheDocument();
  });

  it('should handle adding job after deleting all jobs', async () => {
    render(<Jobselector />);
    
    // Delete all jobs first
    const deleteButtons = screen.getAllByRole('button').filter(button => 
      button.innerHTML.includes('M6 18L18 6M6 6l12 12')
    );
    
    deleteButtons.forEach(button => {
      fireEvent.click(button);
    });

    await waitFor(() => {
      expect(screen.getByText('No jobs')).toBeInTheDocument();
    });

    // Add a new job
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);
    
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: 'First Job After Delete' } });
    
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(screen.getByText('First Job After Delete')).toBeInTheDocument();
      expect(screen.queryByText('No jobs')).not.toBeInTheDocument();
    });
  });

  it('should handle concurrent edit and add operations', async () => {
    render(<Jobselector />);
    
    // Open edit modal
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      
      // Try to open add modal while edit is open
      const addButton = screen.getByText('Add Job');
      fireEvent.click(addButton);

      // The component allows multiple modals to be open simultaneously
      expect(screen.getByText('Edit Job')).toBeInTheDocument();
      expect(screen.getByText('Create New Job')).toBeInTheDocument();
    }
  });

  it('should handle form validation properly', async () => {
    render(<Jobselector />);
    
    // Open add modal
    const addButton = screen.getByText('Add Job');
    fireEvent.click(addButton);

    // Try to submit without entering anything
    const createButton = screen.getByText('Create');
    fireEvent.click(createButton);

    // Modal should still be open
    expect(screen.getByText('Create New Job')).toBeInTheDocument();

    // Try to submit with only spaces
    const input = screen.getByPlaceholderText('Job Title');
    fireEvent.change(input, { target: { value: '   ' } });
    fireEvent.click(createButton);

    // Modal should still be open
    expect(screen.getByText('Create New Job')).toBeInTheDocument();
  });

  it('should handle edit form validation properly', async () => {
    render(<Jobselector />);
    
    // Open edit modal
    const editButtons = screen.getAllByRole('button');
    const editButton = editButtons.find(button => 
      button.innerHTML.includes('M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z')
    );
    
    if (editButton) {
      fireEvent.click(editButton);
      
      // Try to submit with empty value
      const input = screen.getByDisplayValue('Frontend Developer');
      fireEvent.change(input, { target: { value: '' } });
      
      const updateButton = screen.getByText('Update');
      fireEvent.click(updateButton);

      // Modal should still be open
      expect(screen.getByText('Edit Job')).toBeInTheDocument();

      // Try to submit with only spaces
      fireEvent.change(input, { target: { value: '   ' } });
      fireEvent.click(updateButton);

      // Modal should still be open
      expect(screen.getByText('Edit Job')).toBeInTheDocument();
    }
  });
}); 