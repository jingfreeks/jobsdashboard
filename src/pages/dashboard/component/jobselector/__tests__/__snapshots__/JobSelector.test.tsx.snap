// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JobSelector > should render correctly and match snapshot 1`] = `
<div>
  <div
    class="max-w-2xl mx-auto"
  >
    <div
      class="flex items-center justify-between mb-6"
    >
      <h2
        class="text-2xl font-bold text-slate-800"
      >
        Jobs List
      </h2>
      <button
        class="flex items-center gap-2 bg-blue-600 hover:bg-blue-700 text-white px-4 py-2 rounded-lg font-semibold shadow transition"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-circle-plus w-5 h-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M8 12h8"
          />
          <path
            d="M12 8v8"
          />
        </svg>
         Add Job
      </button>
    </div>
    <div
      class="bg-white rounded-xl shadow p-6 border border-slate-100"
    >
      <ul
        class="divide-y divide-slate-100"
      >
        <li
          class="flex items-center justify-between py-3"
        >
          <span
            class="flex-1 truncate text-slate-800 font-medium"
          >
            Frontend Developer
          </span>
          <div
            class="flex gap-2"
          >
            <button
              class="text-blue-500 hover:text-blue-700 px-2 py-1 rounded transition"
            >
              <svg
                class="h-4 w-4"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
            <button
              class="text-red-500 hover:text-red-700 px-2 py-1 rounded transition"
            >
              <svg
                class="h-4 w-4"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 18L18 6M6 6l12 12"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
        </li>
        <li
          class="flex items-center justify-between py-3"
        >
          <span
            class="flex-1 truncate text-slate-800 font-medium"
          >
            Backend Engineer
          </span>
          <div
            class="flex gap-2"
          >
            <button
              class="text-blue-500 hover:text-blue-700 px-2 py-1 rounded transition"
            >
              <svg
                class="h-4 w-4"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
            <button
              class="text-red-500 hover:text-red-700 px-2 py-1 rounded transition"
            >
              <svg
                class="h-4 w-4"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 18L18 6M6 6l12 12"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
        </li>
        <li
          class="flex items-center justify-between py-3"
        >
          <span
            class="flex-1 truncate text-slate-800 font-medium"
          >
            Product Manager
          </span>
          <div
            class="flex gap-2"
          >
            <button
              class="text-blue-500 hover:text-blue-700 px-2 py-1 rounded transition"
            >
              <svg
                class="h-4 w-4"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15.232 5.232l3.536 3.536M9 13h3l8-8a2.828 2.828 0 00-4-4l-8 8v3z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
            <button
              class="text-red-500 hover:text-red-700 px-2 py-1 rounded transition"
            >
              <svg
                class="h-4 w-4"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 18L18 6M6 6l12 12"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
